lib/modules: Short-circuit unmatchedDefns when configs is empty

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Improve performance.

By short-circuiting, we don't have to evaluate as much of the
`options` trees.

Allocation stats show -0.55% percent in heap size.
`bench` shows

Before:

```
benchmarking nix-instantiate -A nixosTests.specialisation -A nixosTests.installer -A nixosTests.nginx -A nixosTests.postgresql
time                 122.2 s    (118.7 s .. 124.9 s, ci 0.990)
                     1.000 R²   (1.000 R² .. 1.000 R², ci 0.990)
mean                 120.8 s    (119.2 s .. 121.8 s, ci 0.990)
std dev              1.170 s    (310.6 ms .. 1.601 s, ci 0.990)
variance introduced by outliers: 19% (moderately inflated)
```

After:

```
time                 121.9 s    (116.2 s .. 128.3 s, ci 0.990)
                     1.000 R²   (NaN R² .. 1.000 R², ci 0.990)
mean                 120.2 s    (118.7 s .. 121.4 s, ci 0.990)
std dev              1.340 s    (313.6 ms .. 1.656 s, ci 0.990)
variance introduced by outliers: 19% (moderately inflated)
```

suggesting a 1.6% performance improvement, although that number is still bogus, considering the possible range of each combined measurement. Maybe I should have picked a lower confidence interval. I was hoping it would produce more samples for a high ci. I might still try that.

Performance report for d6ebd53 https://github.com/NixOS/nixpkgs/pull/144022/checks?check_run_id=4062089224

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

roberth committed:
lib/modules: Short-circuit unmatchedDefns when configs is empty

roberth commented:
@Infinisil could you double-check?

Infinisil commented on lib/modules.nix:
```diff
@@ -430,10 +430,16 @@ rec {
-        mapAttrs (n: v: v.unmatchedDefns) resultsByName
-        # Plus the definitions for the current prefix that don't have a matching option
-        // removeAttrs defnsByName' (attrNames matchedOptions);
+        if configs == []
```
I think this could even be lifted up one level to the `unmatchedDefns` definition

roberth replied:
Done. Will save another few thunks and such.

Infinisil commented on lib/modules.nix:
```diff
@@ -430,10 +430,16 @@ rec {
+          # Propagate all unmatched definitions from nested option sets
+          mapAttrs (n: v: v.unmatchedDefns) resultsByName
+          # Plus the definitions for the current prefix that don't have a matching option
+          // removeAttrs defnsByName' (attrNames matchedOptions);
```
After taking a closer look, I think the main savings come from `defnsByName'` not being evaluated when `configs == []`, meaning that `config` sections aren't evaluated when not needed.

Since this does make module more lazy to some degree it would be awesome if we could have a test case that shows that ensures that extra lazy behavior so we can see the impact and prevent future regressions of it.

roberth replied:
I've tested this for options now, but not for config yet.

roberth committed:
lib/modules: Short-circuit unmatchedDefns earlier

roberth committed:
lib/tests/modules: Test non-strictness some more

Doesn't seem to have been a problem actually, but now it won't
regress.

roberth committed:
lib/modules: Remove a lib.flip

In hot code, the overhead (envs, applies) can matter.

roberth committed:
lib/modules: Fix import* comments

Very confusing otherwise.

roberth committed:
lib/modules: Use strict fold' as recursiveUpdate is also strict

recursiveUpdate does not produce an attrset until it has evaluated
both its arguments to weak head normal form.

    nix-repl> lib.recursiveUpdate (throw "a") (throw "b")
    error: b

    nix-repl> lib.recursiveUpdate (throw "a") {}
    error: a

roberth commented:
I've spotted some low hanging fruit that I've added here because it's simple enough. Will make separate PRs for other improvements though, because ofborg stats aren't comparable as it merges with (mutable) master before each run.

roberth commented:
@Infinisil  That's all, as far as I'm concerned.

There's potential for more laziness, allowing `mkOption` to be replaced by bottom, by removing the `count` call and making `unmatchedDefns` tree-shaped, but I'm not going there with this PR. It seems rather too invasive because that data structure also affects types.

infinisil reviewed (approved):
Nice, looking good!

andir commented:
I haven't invested much time but apparently this broke the way `extraArgs` are passed to module. My `morph`  deployments now don't get the `name` attribute passed to modules anymore and thus I am unable to extract the current nodes name in a module (without resorting to other module system parameters).  It looks like @eadwu ran into the same issue but simply reverted it locally?

roberth commented:
@andir Are you sure it isn't fixed by https://github.com/NixOS/nixpkgs/pull/148315?

I don't know about eadwu's https://github.com/NixOS/nixpkgs/pull/144022#ref-commit-b2c6bf5, but it doesn't seem to be part of any ref anymore.

andir commented:
> @andir Are you sure it isn't fixed by https://github.com/NixOS/nixpkgs/pull/148315?
> 
> I don't know about eadwu's https://github.com/NixOS/nixpkgs/pull/144022#ref-commit-b2c6bf5, but it doesn't seem to be part of any ref anymore.

I think you are right. It must be that other PR. I somehow didn't find that.
